src/python_testing/TC_DeviceBasicComposition.py

```diff
@@ -840,6 +843,129 @@ def test_DESC_2_2(self):
+                        success = False
+                for feature_mask, xml_feature in clusters[cluster_id].features.items():
+                    conformance_decision = xml_feature.conformance(feature_map, attribute_list, all_command_list)
+                    if conformance_decision == ConformanceDecision.MANDATORY and feature_mask not in feature_masks:
+                        self.record_error(self.get_test_name(), location=location,
+                                          problem=f'Required feature with mask 0x{f:02x} is not present in feature map')
+                        success = False
+
+                # Attribute conformance checking
+                for attribute_id, attribute in cluster.items():
+                    location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+                    if attribute_id not in clusters[cluster_id].attributes.keys():
+                        # TODO: Consolidate the range checks with IDM-10.1 once that lands
+                        if attribute_id <= 0x4FFF:
+                            # manufacturer attribute
+                            self.record_error(self.get_test_name(), location=location,
+                                              problem='Standard attribute found on device, but not in spec')
+                            success = False
+                        continue
+                    xml_attribute = clusters[cluster_id].attributes[attribute_id]
+                    conformance_decision = xml_attribute.conformance(feature_map, attribute_list, all_command_list)
+                    if not conformance_allowed(conformance_decision, allow_provisional):
+                        location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+                        self.record_error(self.get_test_name(), location=location,
+                                          problem=f'Attribute 0x{attribute_id:02x} is included, but is disallowed by conformance')
+                        success = False
+                for attribute_id, xml_attribute in clusters[cluster_id].attributes.items():
+                    conformance_decision = xml_attribute.conformance(feature_map, attribute_list, all_command_list)
+                    if conformance_decision == ConformanceDecision.MANDATORY and attribute_id not in cluster.keys():
+                        location = AttributePathLocation(endpoint_id=endpoint_id, cluster_id=cluster_id, attribute_id=attribute_id)
+                        self.record_error(self.get_test_name(), location=location,
+                                          problem=f'Attribute 0x{attribute_id:02x} is required, but is not present on the DUT')
+                        success = False
+
+                def check_spec_conformance_for_commands(self, command_type: CommandType) -> bool:
```

Comment:
This inline method with "self" parameter feels odd. How about moving it as a stand-alone method (and prefix with `_` or `__` if we want it private?